### PR TITLE
Move critical file watching in-house.

### PR DESCRIPTION
**Pull request by olafurpg**
This PR is the first step towards #338.

Previously, we used a different file watching solution for editors that support
file watching and editors with no file watching support.  This commit moves all
file watching to use the same in-house file watching solution for all editors.
The benefit of this change is that we have a single implemementation that we
know works the same for all editors.

However, we still use the editor file watching support (if available)
for nice-to-have but non-critical patterns like
`$workspace/project/*.sbt`. Our in-house implementation only supports
recursive file watching so it's not easy to watch those patterns
ourselves.

Sidenote: the LSP spec does not specify the syntax for targeting precise
directories for dynamic file watching registration. My experiments show
that the syntax supported by VSCode requires is weird on Windows, it
uses backward slashes until the workspace directory, and forward slashes
after that:
```
C:\Users\Bar\workspace/project/*.scala
```

### CODE REVIEW

**Comment by olafurpg** [comment]
FYI @laughedelic ide-scala will need to update the file watching patterns to remove `*.{scala,java}`

The exact patterns are documented now on the website after this PR

```json
{
  "watchers": [
    {
      "globPattern": "file:///to/workspace/*.sbt"
    },
    {
      "globPattern": "file:///to/workspace/project/*.{scala,sbt}"
    },
    {
      "globPattern": "file:///to/workspace/project/build.properties"
    }
  ]
}
```

**Comment by gabro** [comment]
> The benefit of this change is that we have a single implemementation that we
know works the same for all editors.

What were the issues of relying on editors for this? Unreliable detection of changes?

**Comment by olafurpg** [comment]
@gabro it's primarily the lack of a solid spec for how to target precise directories. LSP is silent on the topic and my experiments with VSCode show that the syntax to target directories is super weird

screenshot 2018-12-18 at 17 05 01

We can't confidently test this aspect because it relies on the editor so I feel more confident taking control over it.

**Comment by olafurpg** [comment]
For find references, we will need to target specific directories `META-INF/semanticdb`. I also expect the number of file events to significantly increase with `*.semanticdb` changes so prefer a native solution over doing it via LSP.

File: metals/src/main/scala/scala/meta/internal/metals/FileEvents.scala [hunk 1]
```diff
@@ -1,77 +1,84 @@
...
-        val fileEvent = new FileEvent(uri, changeType)
-        val params = new DidChangeWatchedFilesParams(
-          Collections.singletonList(fileEvent)
-        )
-        didChangeWatchedFiles(params)
+
+  def restart(): Unit = {
+    val paths = new util.ArrayList[Path]()
+    buildTargets.sourceDirectories.foreach { dir =>
+      if (!dir.isDirectory) {
```
Comment (gabro) [inline metals/src/main/scala/scala/meta/internal/metals/FileEvents.scala]: the check is redundant

File: metals/src/main/scala/scala/meta/internal/metals/FileEvents.scala [hunk 2]
```diff
@@ -1,77 +1,84 @@
...
+ * the structure of the build. We don't use the LSP dynamic file watcher capability because
+ *
+ * 1. the glob syntax is not defined in the LSP spec making it difficult to deliver a
+ *    consistent file watching experience with all editor clients on all operating systems.
+ * 2. we may have a lot of file watching events and it's presumably less overhead to
+ *    get the notifications directly from the OS instead of through the editor via LSP.
+ *
+ * Given we rely on file watching for critical functionality like Goto Definition and it's
+ * really difficult to reproduce/investigate file watching issues, I think it's best to
+ * have a single file watching solution that we have control over.
```
Comment (laughedelic) [inline metals/src/main/scala/scala/meta/internal/metals/FileEvents.scala]: 👍

File: docs/editors/new-editor.md [hunk 3]
```diff
@@ -88,16 +88,23 @@ Possible values:
 
 ### `-Dmetals.file-watcher`
 
+This option is no longer used by Metals.
+
+### `-Dmetals.directory-glob`
```
Comment (laughedelic) [inline docs/editors/new-editor.md]: Bikeshedding: IMO the name is not very clear. How about something like `glob-syntax` or `watch-pattern-type` ?

**Pull request merged**